### DISCUSSION TITLE
modbus RTU server - how to define the start-register?

### DISCUSSION THREAD

**Discussion by chrgraf** [eModbus/eModbus#306] [Q&A] [answered]
Dear all,

kindly seek for help.

have on a single ESP32 und use RTU-client function to query 38 registers from a smartmeter starting from register 243.

The derived data from the query  shall be used by another RTU Server instance on the same ESP32.
However (at least how my sketch behaves) the emodbus RTU server instance presents the read data starting from register zero, instead of starting at register 243

What needs to be done to tell the server that the datastore shall be presented with an offset (in mine case offset  243)?

Mostly relying on one of the examples provided by this library.

My FC03 function is looking like this here (not sure if this would be the place to set the offset?)

```
..
#define NUM_VALUES_range1 38
...
void setup() {
//Server-function - connected to Solis inverter
// Register served function code worker for server 1, FC 0x03
MBserver.registerWorker(0x01, READ_HOLD_REGISTER, &FC03);

// Start ModbusRTU background task
  MBserver.begin(Serial2);
}
...

// FC03: worker do serve Modbus function code 0x03 (READ_HOLD_REGISTER)
ModbusMessage FC03(ModbusMessage request) {
  uint16_t address;           // requested register address
  uint16_t words;             // requested number of registers
  ModbusMessage response;     // response message to be sent back

  // get request values
  request.get(2, address);
  request.get(4, words);
  Serial.printf ("address %u  words %u \n",(unsigned) address, (unsigned) words);
  // Address and words valid? We assume 38 registers here for demo
  if (address && words && (address + words) <= NUM_VALUES_all) {
    // Looks okay. Set up message with serverID, FC and length of data
    response.add(request.getServerID(), request.getFunctionCode(), (uint8_t)(words * 2));
    // Fill response with requested data
    for (uint16_t i = address; i < address + words; ++i) {
      response.add(values[i]);                          <<<< values[i] is the datastore
  } else {
    // No, either address or words are outside the limits. Set up error response.
    response.setError(request.getServerID(), request.getFunctionCode(), ILLEGAL_DATA_ADDRESS);
  }
  return response;
}
```

many thanks

christian

**Comment by Miq1** [comment, marked as answer]
The addresses are all virtual in Modbus.How a certain requested address is represented is completely a matter of the server's internal logic. All the examples are using a memory model with registers starting at address 0 and that is the reason for the validity checks all using the 0. But you are free to do it differently - if you want the addressesto start at 243, you will need to adapt the code accordingly.

**Reply by chrgraf** [reply]
Dear Miq1,

thanks alot for clarification.

I am now just populating my datastore response.add(values[i]);  as shown in my above snippet from zero as you say and count beyond my magic 243..

Its doing fine...

I switched  library now to emodbus. Before (by using another modbus impIementation) I  was only able to read 20 Registers from mine smartmeter without data-loss. Now  Ican read all required 38Registers in a single shot! Excellent stuff your library.

 thanks for this excellent stuff
christian

**Reply by Miq1** [reply]
👍🏻
